File: Pragma_Examples/OpenMP/Fortran/8_jacobi/1_jacobi_usm/update.f90
module update_mod
  use kind_mod
  use mesh_mod, only: mesh_t
  implicit none
  !$omp requires unified_shared_memory

  private
  public :: update

contains

  subroutine update(mesh,rhs,au,u,res)
    type(mesh_t), intent(inout) :: mesh
    real(RK), intent(inout) :: rhs(:,:), au(:,:)
    real(RK), intent(inout) :: u(:,:)
    real(RK), intent(inout) :: res(:,:)
    integer(IK) :: i,j
    real(RK) :: temp,factor

    factor = (2._RK/mesh%dx**2+2._RK/mesh%dy**2)**-1

    !$omp target teams distribute parallel do collapse(2) private(temp)
    do j = 1,mesh%n_y
      do i = 1,mesh%n_x
        temp = rhs(i,j) - au(i,j)
        res(i,j) = temp
        u(i,j) = u(i,j) + temp*factor
      end do
    end do
  end subroutine

end module update_mod

File: Pragma_Examples/OpenMP/CXX/8_jacobi/1_jacobi_usm/Update.cpp
//**************************************************************************
//* Copyright (c) 2023, Advanced Micro Devices, Inc. All rights reserved.
//**************************************************************************

#include "Jacobi.hpp"

#pragma omp requires unified_shared_memory
//Jacobi iterative method
// U = U + D^{-1}*(RHS - AU)
void Update(mesh_t& mesh,
            const dfloat factor,
            dfloat* RHS,
            dfloat* AU,
            dfloat* RES,
            dfloat* U) 
{
  const int N = mesh.N;
  #pragma omp target teams distribute parallel for
  for (int id=0;id<N;id++) 
  {
    const dfloat r_res = RHS[id] - AU[id];
    RES[id] = r_res;
    U[id] += r_res*factor;
  }
}
